alerts/apps_alerts.libsonnet

```diff
@@ -235,9 +235,11 @@
             'for': '15m',
           },
           {
-            alert: 'KubeJobCompletion',
+            alert: 'KubeJobNotCompleted',
             expr: |||
-              kube_job_spec_completions{%(prefixedNamespaceSelector)s%(kubeStateMetricsSelector)s} - kube_job_status_succeeded{%(prefixedNamespaceSelector)s%(kubeStateMetricsSelector)s}  > 0
+              time() - kube_job_status_start_time{%(prefixedNamespaceSelector)s%(kubeStateMetricsSelector)s}
+                and
+              kube_job_status_active{%(prefixedNamespaceSelector)s%(kubeStateMetricsSelector)s} > 0
             ||| % $._config,
             'for': '12h',
```

Comment:
Using `time - kube_job_status_start_time > 0` with `for: 12h` assumes that the backing PromQL engine has gapless data for a very long time.  How about:

```
time() - max by (namespace,job_name) (
  kube_job_status_start_time
  and
  (kube_job_status_active > 0)
) > 12*60*60
```

(with whatever grouping we need to get a `namespace,job_name` match in the `and` and also filters for `%(prefixedNamespaceSelector)s%(kubeStateMetricsSelector)`) to say "this job has been active for `$DURATION`"?  With that approach, the duration would be served out of kube-state-metrics, which in turn gets the start time from the Kube API, which is getting data out of etcd.  All current-data stuff, and you wouldn't need to use `for` and rely on hours of gapless Prom data.

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=2057967#c0

Reply:
Thanks @wking , Updated the PR. Please check.

Reply:
Since this is using `max()` and thus selecting only the latest started job I wonder what will be the outcome when the incomplete jobs start to overlap?

Reply:
@paulfantom,  Only one job can exist at time for a given (namespace, job_name) pair. When CronJob creates a Job, it ensures that each instance gets unique name.

Reply:
Your current tip commit is still using `for: 12h`, and the point I'm trying to make is that we don't need (or want) to set `for` at all.  Instead, the cook time should go in the `expr`'s `> 12*60*60`, so we avoid relying on gapless Prom data.

Reply:
phrased another way, etcd has all the data this alarm needs (when the job was created, and the fact that it is still active) to start firing immediately if Prom was just installed on this cluster and just completed its first metrics scrape.  There is no need for cooking on the Prom side here.